src/agent/coverage/src/allowlist.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use anyhow::Result;
+use regex::Regex;
+
+#[derive(Clone, Debug, Default)]
+pub struct TargetAllowList {
+    pub functions: AllowList,
+    pub modules: AllowList,
+    pub source_files: AllowList,
+}
+
+impl TargetAllowList {
+    pub fn new(modules: AllowList, source_files: AllowList) -> Self {
+        // Allow all.
+        let functions = AllowList::default();
+
+        Self {
+            functions,
+            modules,
+            source_files,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct AllowList {
+    allow: Vec<Regex>,
+    deny: Vec<Regex>,
+}
+
+impl AllowList {
+    pub fn is_allowed(&self, path: impl AsRef<str>) -> bool {
+        let path = path.as_ref();
+
+        match (self.allow.is_empty(), self.deny.is_empty()) {
+            (false, false) => {
+                // Allow only if rule-allowed but not also rule-denied.
+                self.has_allow_match(path) && !self.has_deny_match(path)
+            }
+            (false, true) => {
+                // Deny unless rule-allowed.
+                self.has_allow_match(path)
+            }
+            (true, false) => {
+                // Allow unless rule-denied.
+                !self.has_deny_match(path)
+            }
+            (true, true) => {
+                // Allow all.
+                true
+            }
+        }
+    }
+
+    fn has_allow_match(&self, path: &str) -> bool {
+        for re in &self.allow {
```

Comment:
Suggest using [`RegexSet`](https://docs.rs/regex/latest/regex/struct.RegexSet.html) instead?

Reply:
The only reason I didn't is because it felt wasteful to discard the `Regex` values we construct when parsing `AllowListLine`'s, and I wanted to avoid fallibility in the `RegexSet` ctor.

But on a second look, we _can_ easily access the raw regex with `as_str()`, and obtaining those strings (only) from a parsed `Regex` lets us safely `unwrap()` the ctor call.

Reply:
https://github.com/microsoft/onefuzz/pull/2701/commits/71bef6c1cea3650e33a8af83f86f460729205442

Reply:
(adding tests for this)